### PR TITLE
Stop looking at binding patterns for type argument inference

### LINKED ISSUE #43605: Presence of a binding pattern changes inference for callback function parameters
# Bug Report

### 🕗 Version & Regression Information

This bugs occurs in typescript 4.2.3, 4.2.4 and 4.3.0-beta, I didn't test the older typescripts.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.0-beta#code/PTAEBcE8AcFNQCawGYEsB2ryoPboM6j6oC20ANqmrAqMgE44mgAGAAvTQK4AeAVvmDgcOcgGssLADStOAQwDG4FqDnpaLeUoC0nBLxYBYAFAmocUABFU+aHPAKAFgB4AgqAC8oAN4QYsAC5VdEgAblAAbVgecHo5AAVGaHwg-FiMAHMAXSC1SFAAXwA+Tx9QZwAVUGjwWHVCVyKACkVsPCCKgEoOwtCTJAVyOU46LnQlXHRQLnxYAFlYEhxK5uRWnHpIIKbOzxKKmSRkoIAlWDkEPHJIV3o4yGc8koAfafUUDBpu0Aq+4wGhiNkGMJnhprNrLZ7E5KpC7A5HKU4dCXE9mt8KsiEX8TCBQABJSywNL0LhKLj0TKgABkVmJsTJ4ApmRMZn8BIAYiDCF5fPJLuhrpExLAtkR0uhstsAHSy6C5EK7DwlABuOFQtAKOOM5nghPppPJlMllWqMTqCEI+K543wJV5oH5V3yERF+QwoDdOGQPyyAH4ZXKgvFhnISLBavR8JVXaKskUlar1Zrtbq6STGcyTVUaharTaFPgZABVM21eoEokZo2ZFb20BNBA2eFOIJYmFPGTA20dROgYvavHd0EEVnGBR4NKjW0ARlK3hMoCX04Urm2ciCACMRORzuhvmqNXsfAUpIvl8OAELbTepCUZGSb2-i40ZA-J4-eU-npfDgDC2wKEE6BcCQm6wPQMgKEBoAgWBEFQdBW47nu75HsqJ5nsYWomBOBDgCu+AAEzzj+K6WNsCDIaIqFBIetAYV+WHLiuACi2yBC+mQyLAnEkpkaEMSUTFkcOHLbMgwGgeBkF0JJsHSQhcnyduNFqIJn7fthqbshUBb4HOXi6t6hEzn8aZ6dyJFGf4JnDsRg5gFoBGOCIYj4GOw5tFMMywGc+g8O2jgzqauYVvq1ZMq+lT6TORTNEgkVZhkbYGpm0WWbOMgVAmPQLsYLGcFFPmzAsSxNDsx5NL4srSolDI1pKTS+UFOxdtyM67AUnQyBEWSdH8BSeSC3ngn53CBc2KJEZUc5hZalZpY1GQxR1RTZSR81WlWDVRbWmWFkR8VNGR9WGntkozqlSUZbF2Vxcxy5nelmREddu3JattpERt60mBic60hUJH5YVEYUiV8yLDgFV9tVoC1c9y0zs1EJTQibWmT1COykjF0ZERqOwK12P2URXXY31A0mENpjGHitRTtEYYUMSY6+f5vBBSjZGNl2cO+MOa4NrV0B9ggTTINKQtNKLnSkyCAEi3K4uS9LiuyyrFP-dqeFTgA+qUHMTdzJ0FcufN0ALK7C5r0pi8eEtSzLcsK+M17K-bqvOyCl522L2vGNTdN6wRgsgq47Xu1HCh-oURuzJzk1QgiPPm0ulvINbQuBl7jtq1e-vyzHHtF-nPvu0Xbux7nDsYU76vjH+VeFJ0Ovs4nJvo04hO8wg-NVeH7u197jcKH7rsx0rZf1wXGuu63j0Z-3VuD+RI-l2PlhVzH4me3XJQN2JLfdVhwe4ZOYc20E+tC1PN--jHrG7-HXjGwFQW9+nDYr1na85-vUeLstb30AZvf8J8ep9wHhheGw52JgNnhXBQrEd4rj3jPQ+c9xgckgWfP4QA)

### 💻 Code

```ts
// type definitions simplified from `@reduxjs/toolkit`, `react` and `react-redux`

type Dispatch<A = { type: any; [extraProps: string]: any }> = { <T extends A>(action: T): T };
declare function useMemo<T>(factory: () => T, deps: ReadonlyArray<any> | undefined): T;
declare function useDispatch<TDispatch = Dispatch<any>>(): TDispatch;

// IDestructuring & Destructuring

type IFuncs = { readonly [key: string]: (...p: any) => void };

type IDestructuring<T extends IFuncs> = { readonly [key in keyof T]?: (...p: Parameters<T[key]>) => void };

type Destructuring<T extends IFuncs, U extends IDestructuring<T>> = (dispatch: Dispatch<any>, funcs: T) => U;

// functions

const funcs1 = {
    funcA: (a: boolean): void => {},
    funcB: (b: string, bb: string): void => {},
    funcC: (c: number, cc: number, ccc: boolean): void => {},
};
const funcs2 = {
    funcD: (d: boolean): void => {},
    funcE: (e: string, ee: string): void => {},
    funcF: (f: number, ff: number, fff: boolean): void => {},
};

type TFuncs1 = typeof funcs1;
type TFuncs2 = typeof funcs2;

// react hooks

function useReduxDispatch1<T extends IDestructuring<TFuncs1>>(destructuring: Destructuring<TFuncs1, T>): T {
    return useMemo(() => ({ ...destructuring(useDispatch(), funcs1) }), []);
}

function useReduxDispatch2<T1 extends IDestructuring<TFuncs1>, T2 extends IDestructuring<TFuncs2>>(
    destructuring1: Destructuring<TFuncs1, T1>,
    destructuring2: Destructuring<TFuncs2, T2>,
): T1 & T2 {
    return useMemo(() => ({ ...destructuring1(useDispatch(), funcs1), ...destructuring2(useDispatch(), funcs2) }), []);
}

// test examples

useReduxDispatch1(
    (d, f) => ({ funcA: (...p) => d(f.funcA(...p)), funcC: (...p) => d(f.funcC(...p)) })
);

const _ = useReduxDispatch1(
    (d, f) => ({ funcA: (...p) => d(f.funcA(...p)), funcB: (...p) => d(f.funcB(...p)) })
);

const { funcA, funcB, funcC } = useReduxDispatch1(
    (d, f) => ({ funcA: (...p) => d(f.funcB(...p)), funcB: (...p) => d(f.funcB(...p)), funcC: (...p) => d(f.funcC(...p)) })
);

useReduxDispatch2(
    (d, f) => ({ funcB: (...p) => d(f.funcB(...p)), funcC: (...p) => d(f.funcC(...p)) }),
    (d, f) => ({ funcD: (...p) => d(f.funcD(...p)), funcF: (...p) => d(f.funcF(...p)) }),
);

const { funcA: _funcA, funcC: _funcC, funcE, funcF } = useReduxDispatch2(
    (d, f) => ({ funcA: (...p) => d(f.funcA(...p)), funcC: (...p) => d(f.funcC(...p)) }),
    (d, f) => ({ funcE: (...p) => d(f.funcE(...p)), funcF: (...p) => d(f.funcF(...p)) }),
);
```

### 🙁 Actual behavior

In the function call on line 55, typescript cannot infer the parameters of array functions `funcA`, `funcB` and `funcC`, actually it thinks the parameters `...args` is `any[]`.

But this bugs doesn't occur in function calls on line 46, 50, 58 and 63.
### 🙂 Expected behavior

No compilation errors.

Typescript should infer the parameters `...args` of array function `funcA`, `funcB` and `funcC` correctly.

### LINKED ISSUE #45663: When using generic pick, object deconstruction does not provide valid keys to select
# Bug Report

### 🔎 Search Terms

Generic pick on object, type part of object, deconstructing of pick shows all keys of object

### 🕗 Version & Regression Information

- This is a crash
- This first occured today when trying to make some generic object picker
- It was tested on TS 4.3.5 and 4.4.2
- It occures both in TypescriptPlayground and Visual Studio Code
- It decrease developer experience as deconstruction can't be used

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.2#code/JYOwLgpgTgZghgYwgAgLJwA4egHgCrIC8ycIAngDTIBKRJ5AfMgN4BQyHyAFHFAOYAuZHgCUQ6gG5WAX1asYAVxAIwwAPYhkGYAgDWOAPJUCEAB6QQAEwDOyXRDJqYyAwy72y1oXgDaAXSo1ACMAKwB+IQMxZAAFHX0jYQYpRWVVDS14-DcPL2F-aPQsaBSlFXVNbT1DY2QzCxs7BycXHIc83wDkYPDIkRZ2TgQNazBMvWghIuwoGpZkHwBpZFBhP0ifPD9kaSZiAH0eoiYDUIgVADoIcChgCGsuQ9CRC6gISwUkLi41Kh8PKgANzgABs-P1CEw2JwYStnO52hdQAgQQpLPcEWQSLZRP01P8HNi1nRgSCpLDOG8wAooJo1OTONIqMxpES4tVEngGCIpIMOFSaXTQsgwuN7FAfs9kEIquKpLJWMMQKN5nAqEEdnRZVwfAByIK6rrMEhCXVwXXq00Gnb9AD0toAFtAUGAnSQFGA1NYFHw+PcxtYHWoFCDLMhA2oAO7IDUaEFkRUjMZHYjavUGo0m5Bmi0xq26m2sHoXODIe3IJ1vFZjYC2SNqKC6UB8LRQNQzeNExwKZAIUjdECd6wQEHnMZBOSsIA)

### 💻 Code

```ts
// We can quickly address your report if:
//  - The code sample is short. Nearly all TypeScript bugs can be demonstrated in 20-30 lines of code!
//  - It doesn't use external libraries. These are often issues with the type definitions rather than TypeScript bugs.
//  - The incorrectness of the behavior is readily apparent from reading the sample.
// Reports are slower to investigate if:
//  - We have to pare too much extraneous code.
//  - We have to clone a large repo and validate that the problem isn't elsewhere.
//  - The sample is confusing or doesn't clearly demonstrate what's wrong.
interface Mapper<T = any, R = any> {
    (arg: T): R;
}

function pick<O, T extends keyof O>(keys: T[], obj?: O): Pick<O, T>;
function pick<T>(keys: T[]): Mapper;
function pick<O, T extends keyof O>(keys: T[], obj?: O) {
    const picker: Mapper<O, { [K in T]: O[T] }> = _obj => Object.entries(_obj).reduce((o, [key, val]) => {
        if (keys.includes(key as T)) o[key as T] = val;
        return o;
    }, {} as Pick<O, T>);

    return obj ? picker(obj) : picker;
}

const { a, b } = pick(['b'], { a: 'a', b: 'b' }) //here the autosuggest should show b only
const obj = pick(['b'], { a: 'a', b: 'b' })
obj.a // here it is working properly as you can only select b
```

### 🙁 Actual behavior

When using the deconstruction construction after pick, I can select b as well as a. But the a is not in object returned from pick.
When first assigning the return value to const, in later code I can only select b, which is expected.

### 🙂 Expected behavior

In deconstruction I should be able to select only keys, that I declared in pick function. The behaviour should be same as in second example, where the return object is first assigned to `obj` var and then refered.

### PR DESCRIPTION
**Pull request by andrewbranch**
Binding patterns are already sketchy enough as contextual types where type argument inference is _not_ involved, but I see no good reason to let them play a part in actual generic inference.

Fixes #45663
Fixes #43605

Related: #39081

### CODE REVIEW

**Comment by andrewbranch** [comment]
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot test this

File: tests/baselines/reference/destructuringTuple.types [hunk 1]
```diff
@@ -23,20 +23,20 @@ declare var receiver: typeof tuple;
 // Repros from #32140
 
 const [oops1] = [1, 2, 3].reduce((accu, el) => accu.concat(el), []);
->oops1 : any
->[1, 2, 3].reduce((accu, el) => accu.concat(el), []) : number
+>oops1 : never
+>[1, 2, 3].reduce((accu, el) => accu.concat(el), []) : never[]
```
Comment (andrewbranch) [inline tests/baselines/reference/destructuringTuple.types:27]: We used to think `[1, 2, 3].reduce((accu, el) => accu.concat(el), [])` is a _number_. You probably need the (unfortunate) overload resolution error in the function body to get that far off track in the first place, but there’s really no reason why destructuring `[oops1]` from the result should have any effect on what we think this type is.

**Comment by andrewbranch** [comment]
@typescript-bot user test this inline
@typescript-bot run dt
@typescript-bot test this

**Review by sandersn — APPROVED** [review]
This sounds reasonable to me, although I'd like a second opinion from @weswigham in case there are any unforeseen consequences.

(Although shipping early in 4.5 is probably enough to turn up those consequences as well.)

**Review by weswigham — APPROVED** [review]
Yeah, I mean - I made the original change to elide them when type parameters had defaults; binding patterns are pretty bad inference sources - probably the worst ones we have.

**Pull request merged**